### ISSUE TITLE
How to support other languages?

### ISSUE THREAD

**Issue by Stefan300381** [OmerTu/GoogleHomeKodi#62]
Hi,

the first tiny setup is working fine for me, but I wonder what will happen as soon as I implement things like:

"Kodi subtitle previous". In IFTTT I can of course translate "Kodi subtitles $" to something like "Kodi Untertitel $" but what about the "actions" like on/off/previous/next? Are they hardcoded?

Same question for playing a specific episode of a tv series.  The IFTTT command just includes the word "episode" but not the word "season" (Kodi play $ episode #)

Stefan

**Comment by p0psicles** [comment]
You can program everything of course. And there are node modules available for using translation maps. But then you still need to find someone who wants to implement it in a clean way.

**Comment by Stefan300381** [comment]
let me rephrase my question. can you point me to the files I need to touch to translate hardcoded commands into German?

**Comment by sveni-lee** [comment]
so far as I understood could change the pharse you set up in IFTTT to German but the wbhook has to be original....

**Comment by p0psicles** [comment]
So your willing to make the changes again, every time there is an update? Where you planning on forking and keeping your own branch?

**Comment by Stefan300381** [comment]
I could write a script to search and replace specific words in specific files :-) I even doubt that I every update as I am very happy with the current feature set.

**Comment by keydon** [comment]
we could add proper i18n capability, but what words do you want to translate?
i mean there is no userinteraction, it just relays commands. or did i miss something?

**Comment by Stefan300381** [comment]
for example if I want to play a tv servies "season 1 episode 2" would be in German "staffel 1 episode 2". Or "subtitles off" is "untertitel aus". The word "untertitel" is part of the IFTTT trigger (easy to translate), but what about "aus"?

**Comment by keydon** [comment]
ah i see your point now.
well some basic translations are certainly possible to support, but it would feel like a hack to me.

the real issue is that we currently have allowed to leak some of the natural language processing into our app. this is bad, because we will never do as good as the big players like google, amazon etc.
we dont event want to that. all our app should do is relay unambiguous commands from such an service to kodi. 
But IFTTT capabilities (at least on the frist sight) are pretty limited on this aspect. I think they have more possibilities for registered developers, but i did not look into that yet.

my goal would be to receive a technical transcript of the users command, like so:
{
 "token": "yX",
 "subtitles": True/False
}

no matter if the user said, "no subs plz" or "weg mit den doofen untertiteln!"

not sure when or how, but i will have an closer look into that issue this month. maybe there is a middle way.

**Comment by p0psicles** [comment]
We could also create a number of 'locale' language files. Each with a map. And depending of the configured locale, use the map for the required keywords, like {on: translated_on}.

The required translations are really limited. But going the route of passing them using your technical transcript, you have to create an ifttt rule for on and off. Does that make sense?

**Comment by keydon** [comment]
yea sure, the locale-file-approch would work. it event might be sufficent and a good solution for this problem.

And yes my approach would currently mean, that there are two iftt rules. 
but still, i didn look into the advanced ifft-api for registred devs, maybe there is something like in dialogflow. where you just define that there are two entities like $setting and $state. you also define that $state has to value-categories (_On_ and _Off_). then you train your model with a bunch of data meaning _Off_ and a bunch meaning _On_. maybe there are even built-in entities taking care of that common use case. Then we can make AI its magic ;)

**Comment by p0psicles** [comment]
Yeah I've also been playing with api.ai. Have created intents and everything. If ifttt can do something similar like that, that would be awesome.

**Comment by keydon** [comment]
maybe we could transition to dialogflow someday, since its more powerful.
i read somewhere here, that there are reasons against it, like you have to trigger the voice-agent before actually beeing able to do commands. but i am not sure if thats true.

https://developers.google.com/actions/discovery/implicit

but i just started with this whole topic, since i got myself a AIY Voice Kit and a google mini :d

**Comment by keydon** [comment]
what i dont like about iftt, that whenever we add new capabilities to this project, nobody can benefit from that. everybody has to undergo the cumbersome process off adding yet another ifttt rule. afaik there isnt even a copy-applet function ;O

**Comment by p0psicles** [comment]
Well that's one the advantages of api.ai. You can export/import. And provide the export (xml) with the app.

**Comment by p0psicles** [comment]
Disadvantage is that it takes allot of effort to use. As you can't publish it. You have to use it in some sort of debug/test mode.

**Comment by keydon** [comment]
why so? dont you think, that our app might get approved someday and be available to everyone? (sure there is a lot of work to until then)
maybe we even can get kodi-teams's blessing if you think there might be a trademark issue.
or we will have to pick a clunky app-name like 'inofficial kodi remote controller' or somthing xd

**Comment by p0psicles** [comment]
I doubt you could get that to work with a private endpoint.

**Comment by keydon** [comment]
well yes, we would have to ditch glitch at some point and maybe official kodi-team will host the relay app. but i see no problem with the private endpoint.
i think phillips hue is quite similar: Google Home -> Google Cloud -> Phillips Cloud -> My Private Phillips Hue Bridge at my home.
